feat(debugger): frame playground

michalkvasnicak opened the pull request:
## Change Summary

This PR adds playground section to debugger where users can test how their frames are rendered.

TODO:

- [ ] - show syntax errors if there are any
- [ ] - add navigation debugger -> playground, playgroud -> debugger
- [ ] - add a way to debug an existing frame, somehow get the JSX definition if possible

## Merge Checklist

<!-- Check the completed and unnecessary tasks below -->

- [ ] PR has a [Changeset](CONTRIBUTING.md)
- [ ] PR includes documentation if necessary
- [ ] PR updates the boilerplates if necessary

stephancill commented on packages/frames.js/src/middleware/images-worker/handler.tsx:
```diff
@@ -1,10 +1,7 @@
 import { ImageResponse } from "@vercel/og";
-import type { ReactElement } from "react";
-import React from "react";
```
Getting `ReferenceError: React is not defined` because of this

michalkvasnicak replied:
where are you getting this error?

michalkvasnicak replied:
tests are passing just fine also building the frames.js.

stephancill replied:
open the images worker example http://localhost:3010/?url=http://localhost:3000/examples/images-worker

![Screenshot 2024-06-04 at 10 31 48](https://github.com/framesjs/frames.js/assets/5469870/b67ed0a6-a39e-4012-bc76-5ce15b000e9a)

michalkvasnicak replied:
hmm interesting

michalkvasnicak replied:
fixed

stephancill commented on packages/debugger/public/frames.js/index.d.ts:
```diff
@@ -0,0 +1,124 @@
+type JsonObject = { [Key in string]: JsonValue } & {
```
i am concerned about this going out of date when a change is made in the other types files. is there a way we can generate this file on build perhaps?

michalkvasnicak replied:
I was trying to find a way, but the issue is that it always produces definition file per js/ts file, we need to merge them to only one file.

michalkvasnicak replied:
Ok I found a way how to do this.

stephancill commented on packages/debugger/app/components/frame-debugger.tsx:
```diff
@@ -599,6 +611,25 @@ export const FrameDebugger = React.forwardRef<
                     FrameImage={FrameImageNext}
                     allowPartialFrame={true}
                   />
+                  {currentFrameStackItem?.status === "done" && (
+                    <WithTooltip tooltip="Open in Playground">
+                      <Link
+                        href={{
+                          pathname: "/playground",
+                          query: {
+                            frame: JSON.stringify(
+                              frameResultToPlaygroundFrame(
+                                currentFrameStackItem.frameResult
+                              )
+                            ),
+                          },
+                        }}
+                        className="absolute top-0 right-0 p-1 m-2 border border-slate-50 rounded-full z-10 shadow"
+                      >
+                        <TerminalIcon size={24} />
+                      </Link>
+                    </WithTooltip>
+                  )}
```
I am worried about the positioning of this button - perhaps below the frame next to the frame URL, or in the action bar to the left or in the debugger could be better options for placements?

michalkvasnicak replied:
Will do

michalkvasnicak replied:
I moved it to actions on the left.

stephancill reviewed (commented):
nice work! please see feedback below

stephancill commented on packages/debugger/app/playground/frame-result-to-playground-frame.ts:
```diff
@@ -0,0 +1,60 @@
+    typeof result.frame.image === "string" &&
+    !result.frame.image.match(/^data:/i)
+  ) {
+    // do not allow data uris here for now, those can be really long. Prefer storing them to session storage instead?
```
perhaps we can transition away from data URIs by adding an image handler on the default frame response handler that gets activated with a special flag

stephancill replied:
opened a PR with this change https://github.com/framesjs/frames.js/pull/415

michalkvasnicak commented:
@stephancill I removed the playground and instead added new tab `Image` to debugger where you can see rendered svg with bounding boxes.

stephancill commented on docs/pages/reference/core/createFrames.mdx:
```diff
@@ -33,21 +33,43 @@ A string that specifies the base path for all relative URLs in the frame definit
 
```
missing docs for `debug` option

michalkvasnicak replied:
Fixed

stephancill commented on packages/debugger/app/components/frame-debugger-image.tsx:
```diff
@@ -0,0 +1,107 @@
+          </p>
+          <p className="mt-1 text-sm text-gray-500">
+            Make sure you are debugging a frame created by Frames.js and{" "}
+            <span className="font-mono bg-slate-100 p-1 border border-slate-200">
+              debug
+            </span>{" "}
```
would be nice to link to the docs here when `debug` is added

michalkvasnicak replied:
I removed this since this doesn't make sense and goes against what you proposed above.

stephancill reviewed (commented):
would it not perhaps make sense to have a toggle somewhere in the UI which lets you toggle the debug view, which would just set debug to true in the image middleware so that it returns ths svg in debug more instead of png? that way we can avoid adding another tab and an entirely separate renderer

michalkvasnicak commented:
@stephancill I added `Image debugging` switch to sidebar, which is enabled if the frame contains debug info with image. Otherwise it is disabled and has an explanation in tooltip.

<img width="1456" alt="image" src="https://github.com/framesjs/frames.js/assets/174716/be003e02-4369-43ec-a691-863cc1bfe196">


I also removed `Image` tab from diagnostics

michalkvasnicak commented:
@stephancill I did not implement svg image since that requires satori, `@vercel/og` is not able to return svg. I think for now it's good enough.

davidfurlong commented:
is there any way to detect debug mode and disable this toggle + set false if its not on?

michalkvasnicak commented:
@davidfurlong there is no easy way to detect debug mode. We can only rely on the data returned from frames app.

michalkvasnicak commented:
<img width="357" alt="image" src="https://github.com/framesjs/frames.js/assets/174716/f68914d8-3b58-4844-bec8-fe60099ad8eb">

@davidfurlong I applied your suggestions.

davidfurlong commented:
lgtm
